### PR TITLE
T120474: Feed jumpiness

### PR DESCRIPTION
**Pull request by bgerstle**
Phab: [T120474](https://phabricator.wikimedia.org/T120474)

---

Current scroll behavior on master when scrolling doing as sections are added: https://vid.me/MVlM
New behavior: https://vid.me/h1Om

The issue was due to a couple things:

- We updated the placeholder image & nib, but not the actual `estimatedRowHeight` value
- The way we were setting items in section controllers was causing extra KVO callbacks, which caused the table to reload each. time. a. cell. was. added.  by the end, our table be all like :rage4: 

So, we just call `self.fetchError = oops` and `self.fetchedItems = yayNewStuff`, and the setters handle the internal state transitions.

Plus, now we have some tests to verify expected KVO notifications! (Hopefully they pass in Travis :pray:, if not, I'll make them Jenkins-only).

### CODE REVIEW

File: Wikipedia/Code/WMFBaseExploreSectionController.m [hunk 1]
```diff
@@ -204,61 +216,39 @@ - (AnyPromise*)fetchDataIgnoreError:(BOOL)ignoreError ignoreCurrentItems:(BOOL)i
...
-    [self setItemsToPlaceholders];
+    [self setItemsToPlaceholdersOrEmpty];
 }
 
 #pragma mark - Utility
 
-- (void)setItemsToPlaceholders {
+- (void)setItemsToPlaceholdersOrEmpty {
     if (![self shouldShowPlaceholderCell]) {
+        self.items = @[];
```
Comment (coreyfloyd) [inline Wikipedia/Code/WMFBaseExploreSectionController.m]: @bgerstle can you somehow indicate why you are doing this. Its not clear why we are setting this to an empty array.

Comment (bgerstle) [inline Wikipedia/Code/WMFBaseExploreSectionController.m]: so, `items` is `nonnull`, and if the section doesn't support placeholders, it should be an empty array. I could either to do it explicitly here, or implicitly by falling through to the logic below and assuming the non-placeholder section has `numberOfPlaceholderCells == 0`.  I'll add a comment explaining the rationale

**Comment by coreyfloyd** [comment]
:+1:

**Comment by coreyfloyd** [comment]
@bgerstle you're my hero

**Comment by montehurd** [comment]
@bgerstle @coreyfloyd seems to have fixed the refresh jump! :)

**Comment by montehurd** [comment]
:+1:

**Comment by bgerstle** [comment]
@coreyfloyd anyone can look cool

**Pull request merged**